спринт 13 домашка. попытка 1

AleksandrRegulus opened the pull request.

AleksandrRegulus committed:
спринт 13 домашка. попытка 1

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/PlayerActivity.kt:
```diff
@@ -0,0 +1,87 @@
+import com.example.playlistmaker.databinding.ActivityPlayerBinding
+import java.io.Serializable
+
+private const val DEFAULT_TIME_STRING = "0:00"
```
Лучше вынести это в `companion object` класса `PlayerActivity`

AleksandrRegulus replied:
Вынес

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/Track.kt:
```diff
@@ -1,12 +1,37 @@
+    fun getTimeFromMillis(): String {
+        val trackTimeInMilliseconds = trackTimeMillis.toLongOrNull()
+        return if (trackTimeInMilliseconds != null) {
+            SimpleDateFormat("mm:ss", Locale.getDefault()).format(trackTimeInMilliseconds)
```
Объект `SimpleDateFormat("mm:ss", Locale.getDefault())` можно создать как свойство класса и воспользоваться lazy инициализацией (объект проинициализируется только при первом доступе к нему):
`private val dateFormat by lazy { SimpleDateFormat("mm:ss", Locale.getDefault()) }`

AleksandrRegulus replied:
Задумку понял, но реализация хромает. Что бы я не делал валится в ошибку

AleksandrRegulus replied:
Я пишу так 
`val timeFormat: String by lazy {
        val trackTimeInMilliseconds = trackTimeMillis.toLongOrNull()
        if (trackTimeInMilliseconds != null) {
            SimpleDateFormat("mm:ss", Locale.getDefault()).format(trackTimeMillis.toLongOrNull())
        } else DEFAULT_TIME_STRING
    }`

ArturNurtdinov replied:
Проблема в том, что этот класс инициализируется gson`ом, а он не использует конструктор, а использует рефлексию. Поэтому не получится заиспользовать `lazy` инициализацию здесь. Вообще я бы перенёс этот код в то место, где необходимо форматировать строку, чтобы не нагружать модель дополнительной логикой.
И там можно воспользоваться и кешированием `SimpleDateFormat` и `lazy` инициализацией
Подробнее про возникшую проблему можешь прочесть тут - https://stackoverflow.com/questions/57601540/kotlin-throw-npe-when-using-lazy-delegate-with-gson

AleksandrRegulus replied:
Ссылку почитал, проблему понял. А вот как решить пока не очень.

AleksandrRegulus replied:
lazy нам нужно что бы вычислить формат времени один раз, но если нет возможности сделать это в конструкторе, то зачем нам lazy потом? Куда сложить его вычисления?

ArturNurtdinov replied:
Вынести логику с форматированием строки в места, где это используется. А data класс оставить голым классом - только поля с данными и никакой логики

AleksandrRegulus replied:
Можно конечно завести поле с дефолтным значением, а в другом месте его заполнить.. Но оно же будет var

ArturNurtdinov replied:
Тебе в data классе не нужно держать никакую логику
Вот как я это вижу например во вьюхолдере:
1) Вьюхолдер внутри себя кеширует объект `SimpleDateFormat`, тут можно и lazy сделать
2) Во время bind во вьюхолдере вызывается функция `format` и результат функции сетится как текст во вью

ArturNurtdinov replied:
А в этом data классе у тебя не будет ни функции `getTimeFromMillis` ни поля `timeFormat`

ArturNurtdinov replied:
> lazy нам нужно что бы вычислить формат времени один раз, но если нет возможности сделать это в конструкторе, то зачем нам lazy потом? Куда сложить его вычисления?

lazy нужен для отложенной инициализации, чтобы у тебя переменная посчиталась не сразу как объект инициализировался, а при первом доступе к этой переменной. Из-за того что объект инициализируется не через конструктор, а через рефлексию с помощью gson - теряется тот инициализатор, который ты объявляешь. Поэтому в рантайме ты получал NPE

AleksandrRegulus replied:
Понял, да так интересно. Но в этом случае нужно будет еще раз повторить эту логику в активити плеера, т.к. там тоже нужно перевести из миллисекунд. Мне кажется вариант с вынесение во вспомогательный класс правильнее.

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/TracksAdapter.kt:
```diff
@@ -1,6 +1,8 @@
 
+import android.content.Context
 import android.view.LayoutInflater
+import android.view.View
```
Лишние импорты

AleksandrRegulus replied:
почистил

ArturNurtdinov commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -0,0 +1,278 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="4dp"
+        android:text="0:30"
```
```suggestion
        tools:text="0:30"
```
Лучше не хардкодить значение текста в лэйауте. Для предпросмотра можно использовать неймспейс `tools`

AleksandrRegulus replied:
Поправил

ArturNurtdinov commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -0,0 +1,278 @@
+        app:layout_constraintEnd_toStartOf="@id/right_guideline"
+        app:layout_constraintStart_toEndOf="@id/left_guideline"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_chainStyle="spread_inside" />
```
`chainStyle` тут кажется лишний

AleksandrRegulus replied:
убрал

ArturNurtdinov commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -0,0 +1,278 @@
+        android:src="@drawable/ic_add_to_playlist"
+        app:layout_constraintBottom_toBottomOf="@id/btn_play"
+        app:layout_constraintEnd_toStartOf="@id/btn_play"
+        app:layout_constraintHorizontal_chainStyle="spread_inside"
```
здесь тоже `chainStyle` лишний

AleksandrRegulus replied:
тоже убрал

ArturNurtdinov commented on app/src/main/res/layout/activity_player.xml:
```diff
@@ -0,0 +1,278 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        app:barrierDirection="end"
+        app:constraint_referenced_tags="durationTag,genreTag,countryTag,yearTag, albumTag" />
```
Круто, что ты используешь здесь барьер, действительно правильное решение

ArturNurtdinov reviewed (commented):
Хорошая работа. Есть одно замечание с вертикальным отступом между названием трека и названием группы:
Как должно быть:
![image](https://github.com/AleksandrRegulus/PlaylistMaker/assets/49075521/a4897169-4aad-4ccb-acdb-e1fd8ce4fb84)
Как на самом деле:
![image](https://github.com/AleksandrRegulus/PlaylistMaker/assets/49075521/be1f73b4-5e60-4c46-8277-f229072bb26a)

AleksandrRegulus committed:
спринт 13 домашка. попытка 2

AleksandrRegulus committed:
спринт 13 домашка. попытка 3

Оставил в покое класс Track.
Вынес логику во вспомогательный класс FormatTrack.
Считаю, что в данном случае lazy совершенно не нужен.
Приложение работает без сбоев.

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/FormatTrack.kt:
```diff
@@ -0,0 +1,27 @@
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+class FormatTrack {
+    companion object {
```
Можно сделать просто вот так:
```suggestion
object FormatTrack {
```

AleksandrRegulus replied:
Да, так еще красивее)

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/FormatTrack.kt:
```diff
@@ -0,0 +1,27 @@
+            } else DEFAULT_TIME_STRING
+        }
+
+        fun getCoverArtwork(artworkUrl100: String?) =
+            if (artworkUrl100.isNullOrEmpty()) "" else artworkUrl100.replaceAfterLast( // получаем ссылку на большой постер
+                '/',
+                "512x512bb.jpg"
+            )
+
+        fun getYearFromReleaseDate(releaseDate: String?) =
+            if (releaseDate.isNullOrEmpty()) "" else releaseDate.substring(0..3)
```
Это можно было оставить в том классе, тут нет никакой логики - просто работа с данными

AleksandrRegulus replied:
Пусть уже все будет в одной куче, а дата класс голый. Мне кажется так читабельнее )
